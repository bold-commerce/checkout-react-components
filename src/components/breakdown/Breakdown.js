import React from 'react';
import { Price, Button } from '@boldcommerce/stacks-ui';
import useBreakdown from '../../hooks/useBreakdown';
import useDiscount from '../../hooks/useDiscount';
import './Breakdown.css';

const Breakdown = () => {
  const {
    subTotal,
    shippingTotal,
    discountTotal,
    total,
    taxes,
    payments,
    taxesIncluded,
    taxesTotal,
    paymentsMade,
    paymentStatus,
    remainingBalance,
  } = useBreakdown();

  const {
    discountApplied,
    discountCode,
    removeDiscount,
  } = useDiscount();

  const paymentMethods = payments.map((payment, index) => (
    <div
      className="Payment__Applied"
      key={
        index
      }
    >
      <div className="Payment__Remove">
        {payment.brand}
        :
        {payment.bin}
      </div>
      <Price amount={
        payment.value
      }
      />
    </div>
  ));

  const readOnly = (paymentsMade && paymentStatus) ?? false;

  return (
    <div className="Summary__Breakdown">
      <div className="SummaryBlock">
        <div className="Summary__Subtotal">
          <div className="SummaryItem__Label">Subtotal</div>
          <div className="SummaryItem__Value">
            <Price amount={subTotal} />
          </div>
        </div>
        <div className="Summary__Discount">
          <div className="Summary__Discount--Labels">
            <div className="SummaryItem__Label" data-test={discountTotal}>Discounts</div>
            {
              !discountApplied && (
                <div className="SummaryItem__Value">$0.00</div>
              )
            }
          </div>
          {
            discountApplied && (
              <div className="Discount__Applied">
                <div className="Discount__Remove">
                  <Button
                    disabled={readOnly}
                    secondary
                    onClick={() => {
                      removeDiscount();
                    }}
                  >
                    X
                  </Button>
                  <div>
                    Discount code:
                    {' '}
                    {discountCode}
                  </div>
                </div>
                <div className="Discount__Amount">
                  <Price amount={discountTotal} />
                </div>
              </div>
            )
          }
        </div>
        <div className="Summary__Shipping">
          <div className="SummaryItem__Label">Shipping</div>
          <div className="SummaryItem__Value">
            <Price amount={shippingTotal} />
          </div>
        </div>
        <div className="Summary__Taxes">
          <div className="SummaryItem__Label">
            Taxes
            {
              taxesIncluded && ' (Included)'
            }
          </div>
          <div className="SummaryItem__Value">
            {
              taxes.length === 0 ? '$0.00'
                : <Price amount={taxesTotal} />
            }
          </div>
        </div>
      </div>
      <div className="SummaryBlock Summary__Total">
        <div className="SummaryItem__Label">Total</div>
        <div className="SummaryItem__Value">
          <Price amount={total} />
        </div>
      </div>
      {readOnly
        && (
          <div className="SummaryBlock Summary__Payments">
            <div className="Summary__Payments--Labels">
              <div className="SummaryItem__Label">Payments</div>
              {!paymentsMade && paymentStatus
              && (
                <div className="SummaryItem__Value">
                  <Price amount={
                    0
                  }
                  />
                </div>
              )}
            </div>
            {
              paymentMethods
            }
          </div>
        )}
      {readOnly
        && (
          <div className="SummaryBlock Summary__Balance">
            <div className="SummaryItem__Label">Remaining Balance</div>
            <div className="SummaryItem__Value">
              <Price amount={
                remainingBalance
              }
              />
            </div>
          </div>
        )}
    </div>
  );
};

export default Breakdown;
