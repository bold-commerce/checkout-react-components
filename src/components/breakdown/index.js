// eslint-disable-next-line import/prefer-default-export
export { default as Breakdown } from './Breakdown';
