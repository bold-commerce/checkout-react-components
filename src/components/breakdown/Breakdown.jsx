/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { Price, Button } from '@boldcommerce/stacks-ui';
import { useDiscount, useBreakdown } from '../../hooks';
import './Breakdown.css';
import { RedactedCreditCard } from '../redacted_credit_card';

const Breakdown = ({
  subTotal, shippingTotal, discountTotal, total, remainingBalance, taxesTotal, payments, hasPayments, taxesIncluded, taxes, discountApplied, discountCode, removeDiscount,
}) => {
  const paymentMethods = payments.map((payment, index) => (
    <div
      className="Payment__Applied"
      key={
        index
      }
    >
      <div className="Payment__Remove">
        <RedactedCreditCard brand={payment.brand} lineText={payment.lineText} />
      </div>
      <Price amount={
        payment.value
      }
      />
    </div>
  ));

  const readOnly = (hasPayments) ?? false;

  return (
    <div className="Summary__Breakdown">
      <div className="SummaryBlock">
        <div className="Summary__Subtotal">
          <div className="SummaryItem__Label">Subtotal</div>
          <div className="SummaryItem__Value">
            <Price amount={subTotal} />
          </div>
        </div>
        <div className="Summary__Discount">
          <div className="Summary__Discount--Labels">
            <div className="SummaryItem__Label" data-test={discountTotal}>Discounts</div>
            {
              !discountApplied && (
                <div className="SummaryItem__Value">$0.00</div>
              )
            }
          </div>
          {
            discountApplied && (
              <div className="Discount__Applied">
                <div className="Discount__Remove">
                  <Button
                    disabled={readOnly}
                    secondary
                    onClick={() => removeDiscount(discountCode)}
                  >
                    X
                  </Button>
                  <div>
                    Discount code:
                    {' '}
                    {discountCode}
                  </div>
                </div>
                <div className="Discount__Amount">
                  <Price amount={discountTotal} />
                </div>
              </div>
            )
          }
        </div>
        <div className="Summary__Shipping">
          <div className="SummaryItem__Label">Shipping</div>
          <div className="SummaryItem__Value">
            <Price amount={shippingTotal} />
          </div>
        </div>
        <div className="Summary__Taxes">
          <div className="SummaryItem__Label">
            Taxes
            {
              taxesIncluded && ' (Included)'
            }
          </div>
          <div className="SummaryItem__Value">
            {
              taxes.length === 0 ? '$0.00'
                : <Price amount={taxesTotal} />
            }
          </div>
        </div>
      </div>
      <div className="SummaryBlock Summary__Total">
        <div className="SummaryItem__Label">Total</div>
        <div className="SummaryItem__Value">
          <Price amount={total} />
        </div>
      </div>
      {readOnly
        && (
          <div className="SummaryBlock Summary__Payments">
            <div className="Summary__Payments--Labels">
              <div className="SummaryItem__Label">Payments</div>
            </div>
            {
              paymentMethods
            }
          </div>
        )}
      {readOnly
        && (
          <div className="SummaryBlock Summary__Balance">
            <div className="SummaryItem__Label">Remaining Balance</div>
            <div className="SummaryItem__Value">
              <Price amount={
                remainingBalance
              }
              />
            </div>
          </div>
        )}
    </div>
  );
};

Breakdown.propTypes = {
  subTotal: PropTypes.number,
  shippingTotal: PropTypes.number,
  discountTotal: PropTypes.number,
  total: PropTypes.number,
  remainingBalance: PropTypes.number,
  taxesTotal: PropTypes.number,
  payments: PropTypes.array,
  hasPayments: PropTypes.bool,
  taxesIncluded: PropTypes.bool,
  taxes: PropTypes.array,
  discountApplied: PropTypes.bool,
  discountCode: PropTypes.string,
  removeDiscount: PropTypes.func,
};

const MemoizedBreakdown = React.memo(Breakdown);

const BreakdownContainer = () => {
  const {
    subTotal, shippingTotal, discountTotal, total, remainingBalance, taxesTotal, payments, hasPayments, taxesIncluded, taxes,
  } = useBreakdown();
  const { discountApplied, discountCode, removeDiscount } = useDiscount();

  return (
    <MemoizedBreakdown
      subTotal={subTotal}
      shippingTotal={shippingTotal}
      discountTotal={discountTotal}
      total={total}
      remainingBalance={remainingBalance}
      taxesTotal={taxesTotal}
      payments={payments}
      hasPayments={hasPayments}
      taxesIncluded={taxesIncluded}
      taxes={taxes}
      discountApplied={discountApplied}
      discountCode={discountCode}
      removeDiscount={removeDiscount}
    />
  );
};

export default BreakdownContainer;
