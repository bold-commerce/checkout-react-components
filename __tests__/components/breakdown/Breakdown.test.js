import React from 'react';
import { render } from '@testing-library/react';
import { Breakdown } from '../../../src/components/breakdown/Breakdown';
import testApplicationState from '../../utils/applicationStateHelper';

const testData = {
  subTotal: 22500,
  shippingTotal: 1000,
  discountTotal: 0,
  total: 26200,
  remainingBalance: 0,
  taxesTotal: 2700,
  payments: testApplicationState.payments,
  hasPayments: true,
  taxesIncluded: false,
  taxes: testApplicationState.taxes,
  discountApplied: false,
  discountCode: '',
};

describe('Breakdown', () => {
  test('renders Breakdown component', () => {
    const { asFragment } = render(
      <Breakdown
        subTotal={testData.subTotal}
        shippingTotal={testData.shippingTotal}
        discountTotal={testData.discountTotal}
        total={testData.total}
        remainingBalance={testData.remainingBalance}
        taxesTotal={testData.taxesTotal}
        payments={testData.payments}
        hasPayments={testData.hasPayments}
        taxesIncluded={testData.taxesIncluded}
        taxes={testData.taxes}
        discountApplied={testData.discountApplied}
        discountCode={testData.discountCode}
        removeDiscount={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Breakdown component with discount applied', () => {
    const { asFragment } = render(
      <Breakdown
        subTotal={testData.subTotal}
        shippingTotal={testData.shippingTotal}
        discountTotal={100}
        total={testData.total}
        remainingBalance={testData.remainingBalance}
        taxesTotal={testData.taxesTotal}
        payments={testData.payments}
        hasPayments={testData.hasPayments}
        taxesIncluded={testData.taxesIncluded}
        taxes={testData.taxes}
        discountApplied
        discountCode="TestDiscount"
        removeDiscount={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Breakdown component with taxes included', () => {
    const { asFragment } = render(
      <Breakdown
        subTotal={testData.subTotal}
        shippingTotal={testData.shippingTotal}
        discountTotal={testData.discountTotal}
        total={testData.total}
        remainingBalance={testData.remainingBalance}
        taxesTotal={testData.taxesTotal}
        payments={testData.payments}
        hasPayments={testData.hasPayments}
        taxesIncluded
        taxes={testData.taxes}
        discountApplied={testData.discountApplied}
        discountCode={testData.discountCode}
        removeDiscount={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
