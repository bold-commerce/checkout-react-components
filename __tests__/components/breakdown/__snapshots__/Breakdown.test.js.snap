// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breakdown renders Breakdown component 1`] = `
<DocumentFragment>
  <div
    class="Summary__Breakdown"
  >
    <div
      class="SummaryBlock"
    >
      <div
        class="Summary__Subtotal"
      >
        <div
          class="SummaryItem__Label"
        >
          Subtotal
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$225.00
            </p>
          </div>
        </div>
      </div>
      <div
        class="Summary__Discount"
      >
        <div
          class="Summary__Discount--Labels"
        >
          <div
            class="SummaryItem__Label"
            data-test="0"
          >
            Discounts
          </div>
          <div
            class="SummaryItem__Value"
          >
            $0.00
          </div>
        </div>
      </div>
      <div
        class="Summary__Shipping"
      >
        <div
          class="SummaryItem__Label"
        >
          Shipping
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$10.00
            </p>
          </div>
        </div>
      </div>
      <div
        class="Summary__Taxes"
      >
        <div
          class="SummaryItem__Label"
        >
          Taxes
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$27.00
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Total"
    >
      <div
        class="SummaryItem__Label"
      >
        Total
      </div>
      <div
        class="SummaryItem__Value"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$262.00
          </p>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Payments"
    >
      <div
        class="Summary__Payments--Labels"
      >
        <div
          class="SummaryItem__Label"
        >
          Payments
        </div>
      </div>
      <div
        class="Payment__Applied"
      >
        <div
          class="Payment__Remove"
        >
          <span
            class="RedactedCard"
          >
            <span
              class="RedactedCard__Brand"
            >
              visa
            </span>
            <span
              class="RedactedCard__Number"
            >
               •••• •••• •••• 1111
            </span>
          </span>
        </div>
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$262.00
          </p>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Balance"
    >
      <div
        class="SummaryItem__Label"
      >
        Remaining Balance
      </div>
      <div
        class="SummaryItem__Value"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$0.00
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Breakdown renders Breakdown component with discount applied 1`] = `
<DocumentFragment>
  <div
    class="Summary__Breakdown"
  >
    <div
      class="SummaryBlock"
    >
      <div
        class="Summary__Subtotal"
      >
        <div
          class="SummaryItem__Label"
        >
          Subtotal
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$225.00
            </p>
          </div>
        </div>
      </div>
      <div
        class="Summary__Discount"
      >
        <div
          class="Summary__Discount--Labels"
        >
          <div
            class="SummaryItem__Label"
            data-test="100"
          >
            Discounts
          </div>
        </div>
        <div
          class="Discount__Applied"
        >
          <div
            class="Discount__Remove"
          >
            <button
              aria-busy="false"
              aria-disabled="true"
              class="stx-button stx-button--secondary stx-button--is-disabled"
              disabled=""
            >
              <span
                class="stx-button__text"
              >
                X
              </span>
              <span
                class="stx-loading-spinner stx-button__spinner"
              />
            </button>
            <div>
              Discount code: TestDiscount
            </div>
          </div>
          <div
            class="Discount__Amount"
          >
            <div
              class="stx-price stx-price--alight-center"
            >
              <p
                aria-label="Product price"
                class="stx-price__amount"
              >
                CA$1.00
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="Summary__Shipping"
      >
        <div
          class="SummaryItem__Label"
        >
          Shipping
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$10.00
            </p>
          </div>
        </div>
      </div>
      <div
        class="Summary__Taxes"
      >
        <div
          class="SummaryItem__Label"
        >
          Taxes
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$27.00
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Total"
    >
      <div
        class="SummaryItem__Label"
      >
        Total
      </div>
      <div
        class="SummaryItem__Value"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$262.00
          </p>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Payments"
    >
      <div
        class="Summary__Payments--Labels"
      >
        <div
          class="SummaryItem__Label"
        >
          Payments
        </div>
      </div>
      <div
        class="Payment__Applied"
      >
        <div
          class="Payment__Remove"
        >
          <span
            class="RedactedCard"
          >
            <span
              class="RedactedCard__Brand"
            >
              visa
            </span>
            <span
              class="RedactedCard__Number"
            >
               •••• •••• •••• 1111
            </span>
          </span>
        </div>
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$262.00
          </p>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Balance"
    >
      <div
        class="SummaryItem__Label"
      >
        Remaining Balance
      </div>
      <div
        class="SummaryItem__Value"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$0.00
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Breakdown renders Breakdown component with taxes included 1`] = `
<DocumentFragment>
  <div
    class="Summary__Breakdown"
  >
    <div
      class="SummaryBlock"
    >
      <div
        class="Summary__Subtotal"
      >
        <div
          class="SummaryItem__Label"
        >
          Subtotal
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$225.00
            </p>
          </div>
        </div>
      </div>
      <div
        class="Summary__Discount"
      >
        <div
          class="Summary__Discount--Labels"
        >
          <div
            class="SummaryItem__Label"
            data-test="0"
          >
            Discounts
          </div>
          <div
            class="SummaryItem__Value"
          >
            $0.00
          </div>
        </div>
      </div>
      <div
        class="Summary__Shipping"
      >
        <div
          class="SummaryItem__Label"
        >
          Shipping
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$10.00
            </p>
          </div>
        </div>
      </div>
      <div
        class="Summary__Taxes"
      >
        <div
          class="SummaryItem__Label"
        >
          Taxes (Included)
        </div>
        <div
          class="SummaryItem__Value"
        >
          <div
            class="stx-price stx-price--alight-center"
          >
            <p
              aria-label="Product price"
              class="stx-price__amount"
            >
              CA$27.00
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Total"
    >
      <div
        class="SummaryItem__Label"
      >
        Total
      </div>
      <div
        class="SummaryItem__Value"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$262.00
          </p>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Payments"
    >
      <div
        class="Summary__Payments--Labels"
      >
        <div
          class="SummaryItem__Label"
        >
          Payments
        </div>
      </div>
      <div
        class="Payment__Applied"
      >
        <div
          class="Payment__Remove"
        >
          <span
            class="RedactedCard"
          >
            <span
              class="RedactedCard__Brand"
            >
              visa
            </span>
            <span
              class="RedactedCard__Number"
            >
               •••• •••• •••• 1111
            </span>
          </span>
        </div>
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$262.00
          </p>
        </div>
      </div>
    </div>
    <div
      class="SummaryBlock Summary__Balance"
    >
      <div
        class="SummaryItem__Label"
      >
        Remaining Balance
      </div>
      <div
        class="SummaryItem__Value"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$0.00
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
